### DISCUSSION TITLE
让最新版xray-core支持在Windows7下运行的方法，还在用老系统的可以试试。

### DISCUSSION THREAD

**Discussion by dashabiguard** [XTLS/Xray-core#3373] [Show and tell]
使用go1.21.4编译，Windows7也可以运行，无需修改代码。
只需修改go.mod，把开头的go 1.22改成go1.21。
然后使用以下命令更新go.mod：
go mod tidy
最后正常编译即可：
go build -o xray.exe -trimpath -ldflags "-s -w -buildid=" ./main

**Comment by RPRX** [comment]
看得我都心动了

**Reply by Einsfier** [reply]
我有点好奇，阴阳怪气是Xray社区的风格吗？

**Reply by uuonda** [reply]
@RPRX Is there a specific reason this is not recommended?

**Reply by RPRX** [reply]
> 我有点好奇，阴阳怪气是Xray社区的风格吗？

？不是，这怎么是阴阳怪气了？我是真的心动了，说实话我是这里唯一一个还想支持 Win7 的开发者，主要是想让 Win7 上的 Xray 能更新到新版本，真的有点想把 v1.9.x 的编译改成 Go 1.21

还有，阴阳怪气是隔壁某两个社区的风格，都人尽皆知的

**Reply by Einsfier** [reply]
那是我有误会了，道个歉：I apologize for that.
受某些不好好说话的人影响，导致戴上了有色眼镜。

**Comment by ErbiGuard** [comment]
> 看得我都心动了

"Please don't be ambiguous. There are still people using Windows 7. You can refuse feedback about Windows 7, and you don't have to release compatible versions. But please don't prevent other users from finding ways to continue supporting Windows 7."

**Reply by RPRX** [reply]
我不知道我就说了个“看得我都心动了”，你们是怎么觉得我反对继续支持 Win7 的？我也在找方法，看一下 https://github.com/XTLS/Xray-core/issues/2942#issuecomment-1925137200

**Comment by ErbiGuard** [comment]
> @RPRX Is there a specific reason this is not recommended?

Please refer to the update log of the Go programming language for details. Use at your own risk, as this project will not accept issue reports related to compiling using this method.

**Comment by chise0713** [comment]
```go
//go:build go1.22
package xx
// ...
```
```go
//go:build !go1.22
package xx
// ...
```
How about we use multiple files for different version of golang?
Then we can disable some features rely on go1.22 when we compile the core using go1.21.

**Reply by sendguard** [reply]
Hope the repository maintainers can consider this suggestion. Additionally, the current version does not require any functionality dependent on go1.22, so there's no need to disable any features. We'll see if future updates necessitate go1.22 functionality. It's suggested to compile three executable files using three different versions of Go, similar to the "MetaCubeX/mihomo" repository (latest Go version, go1.21.4, go1.20.14). For go1.20.14, MultipathTCP needs to be disabled.

**Comment by test4uninstall** [comment]
v1.8.15兼容需要修改代码
打开“transport\internet\httpupgrade\dialer.go”
将里面第266行中的：
```
for _ = range 5 {
```
修改为：
```
for i := 0; i < 5; i++ {
```
然后正常编译即可

**Comment by RPRX** [comment]
https://github.com/XTLS/Xray-core/pull/3455#issuecomment-2222340229

**Reply by test5uninstall** [reply]
> [#3455 (comment)](https://github.com/XTLS/Xray-core/pull/3455#issuecomment-2222340229)

Go1.21.4编译的可以在Win7上运行，无需降到Go1.20。

**Reply by RPRX** [reply]
> Go1.21.4编译的可以在Win7上运行，无需降到Go1.20。

Go 1.21.5 不行吗

**Reply by RPRX** [reply]
> > Go1.21.4编译的可以在Win7上运行，无需降到Go1.20。
> 
> Go 1.21.5 不行吗

看到了 https://github.com/XTLS/Xray-core/pull/3455#issuecomment-2223079850

**Comment by vrnobody** [comment]
golang v1.21.4 的 mac 单元测试会失败，所以 win7 和 mac 不可兼得。还是单独编译个 win7 版发布文件更合适。

**Reply by RPRX** [reply]
我也觉得这是长久之计，毕竟 Go 1.21 快停止维护了，文件名预定 Xray-win7-64.zip 和 Xray-win7-32.zip

**Reply by test5uninstall** [reply]
> 我也觉得这是长久之计，毕竟 Go 1.21 快停止维护了，文件名预定 Xray-win7-64.zip 和 Xray-win7-32.zip

建议用 Xray-windows-go121-64.zip 和 Xray-windows-go121-32.zip

**Reply by test5uninstall** [reply]
go121版本的xray不需要禁用某些功能。

**Reply by RPRX** [reply]
> 建议用 Xray-windows-go121-64.zip 和 Xray-windows-go121-32.zip

有点不太符合现在“Xray-系统-架构.zip”的格式，既然是 Win7 专版就直接写 Win7 吧
